packages/fxa-settings/src/components/UnitRowSecondaryEmail/index.tsx

```diff
@@ -36,21 +57,106 @@ export const UnitRowSecondaryEmail = () => {
-  // needing a mutation - we must have an `onError` option in mutations
-  // to allow tests to pass but providing one prevents an error from actually
-  // throwing so we need to manually report it to Sentry.
-  // See https://github.com/apollographql/react-apollo/issues/2614
-  const [resendSecondaryEmailCode, { data, error }] = useMutation(
-    RESEND_SECONDARY_EMAIL_CODE_MUTATION,
-    {
-      onError: (error) => {
-        sentryMetrics.captureException(error);
-      },
-    }
+  const showSuccess = useCallback(
+    (message: string) => {
+      setAlertMessage([message, 'success']);
+      revealAlertBar();
+    },
+    [setAlertMessage, revealAlertBar]
   );
 
+  const showError = useCallback(
+    (message: string) => {
+      setAlertMessage([message, 'error']);
+      revealAlertBar();
+    },
+    [setAlertMessage, revealAlertBar]
+  );
+
+  const [resendEmailCode] = useHandledMutation(RESEND_EMAIL_CODE_MUTATION, {
+    onCompleted() {
+      showSuccess(
+        `Check the inbox for ${actionableEmail} to verify your secondary email.`
+      );
+    },
+    onError(error) {
+      showError(`Sorry, there was a problem re-sending the verification code.`);
+      throw error;
+    },
+  });
+
+  const [makeEmailPrimary] = useHandledMutation(MAKE_EMAIL_PRIMARY_MUTATION, {
+    onCompleted() {
+      showSuccess(`${actionableEmail} is now your primary email.`);
+    },
+    onError(error) {
+      showError(`Sorry, there was a problem changing your primary email.`);
+      throw error;
+    },
+    update: (cache) => {
+      cache.modify({
+        fields: {
+          account: (existing: Account) => {
+            const emails = existing.emails.map((m) => {
+              if (m.email === actionableEmail) {
+                return { ...m, isPrimary: true };
+              }
+
+              if (m.isPrimary && m.email !== actionableEmail) {
+                return { ...m, isPrimary: false };
+              }
+
+              return { ...m };
+            });
+
+            return { ...existing, emails };
+          },
+        },
+      });
+    },
+  });
+
+  const [deleteEmail] = useHandledMutation(DELETE_EMAIL_MUTATION, {
+    onCompleted() {
+      showSuccess(`${actionableEmail} email successfully deleted.`);
+    },
+    onError(error) {
+      showError(`Sorry, there was a problem deleting this email.`);
+      throw error;
+    },
+    update: (cache) => {
+      cache.modify({
+        fields: {
+          account: (existing: Account) => {
+            const emails = existing.emails
+              .map((m) => {
+                if (m.email === actionableEmail) {
+                  return null;
+                }
+
+                return { ...m };
+              })
+              .filter((m) => !!m);
+
+            return { ...existing, emails };
+          },
+        },
+      });
+    },
+  });
+
+  useEffect(() => {
+    if (queuedAction && session.verified) {
+      queuedAction();
+    }
+  }, [queuedAction, session]);
```

Comment:
these session bit make me feel like there's a new component wanting to be made. I made a change to ModalVerifySession that makes this a bit nicer here: https://github.com/mozilla/fxa/commit/cc97bedbbe56bf92f8f2d77727b40cd2fec1d7fa ... maybe there's a way to encapsulate the action and trigger too. I think there's other areas where we'll need something like this.

Reply:
Oh nice, your change looks good. I think I can build off that.